### PR TITLE
Add Sophia Explained

### PR DESCRIPTION
**Pull request by hanssv**
This is planned to be a place to add more in-depth discussion on Sophia topics...

### CODE REVIEW

File: contracts/sophia_explained.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
+[back](./contracts.md)
```
Comment (dincho) [inline contracts/sophia_explained.md]: double back

Comment (hanssv) [reply contracts/sophia_explained.md]: Copy pasted from `contracts/sophia.md` :joy:

Will fix.

Comment (hanssv) [reply contracts/sophia_explained.md]: Fixed.

**Comment by hanssv** [comment]
> I would like just to make sure if I got it right:
> 
>     * the `data` field is the latest event parameter, which is always string and always non-indexed (what non-indexed means ?)
No, the position is arbitrary it can be in any position, but only one and non-indexed. The `indexed` keyword was a mistake I think,  it made sense in an earlier variant of the design. 
> 
>     * the event parameters are called topics ? Except the last one, non-indexed, string parameter, which is called data?
Yes
> 
>     * the hash of event name/constructor is always the first topic
Yes 
>     * I assume there cannot be same constructor/different parameters (signature) ?
No, this is built-into Sophia itself, a datatype can not have overlapping constructors.
> 
> I find the terminology and fields just a little bit confusing, but otherwise this docs (assuming above is true) explains it well. 👍

**Comment by dincho** [comment]
> No, the position is arbitrary it can be in any position, but only one and non-indexed. The `indexed` keyword was a mistake I think, it made sense in an earlier variant of the design.

OK, so the data is then defined by the non-indexed (without indexed keyword) string param ?

Then if I wrote it as:
```
AnotherEvent(string, indexed address)
```

And then:

```
	  public function emit_events() : () =
	    Chain.event(TheFirstEvent(42))
	    Chain.event(AnotherEvent("This is not indexed", Contract.address))
```

I should get the exactly the same output ?

**Comment by hanssv** [comment]
> > No, the position is arbitrary it can be in any position, but only one and non-indexed. The `indexed` keyword was a mistake I think, it made sense in an earlier variant of the design.
> 
> OK, so the data is then defined by the non-indexed (without indexed keyword) string param ?
> 
> Then if I wrote it as:
> 
> ```
> AnotherEvent(string, indexed address)
> ```
> 
> And then:
> 
> ```
> 	  public function emit_events() : () =
> 	    Chain.event(TheFirstEvent(42))
> 	    Chain.event(AnotherEvent("This is not indexed", Contract.address))
> ```
> 
> I should get the exactly the same output ?

Yes, exactly!

File: contracts/sophia_explained.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                          1875564187002476023854543820981509249331367502514562901623081521315128929137]}],
+  <<"return_type">> => <<"ok">>,
+  <<"return_value">> => <<"cb_AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAArMtts">>}
+```
+
+First look at the `data` field. For the first event it is `<<"cb_Xfbg4g==">>`
+which decodes to `<<>>`, i.e. no data (and that is expected since the event did
+not have any extra data.). For the second event we see
+`<<"cb_VGhpcyBpcyBub3QgaW5kZXhlZK+w140=">>` which decodes to `<<"This is not
+indexed">>`. For the `topics` field we see that there is actually two values
```
Comment (maptuhec) [inline contracts/sophia_explained.md]: I think that the first value of the `topics` field can be explained a bit more. For the first example event the `indexed int` value is presented as a plain integer (42), but for the second event the `indexed address` value is, I assume, hash of the address. The logic behind this is not explained, what would be the value if the parameter is a `boolean`?

**Pull request merged**